PerfTools/AllocMonitor/plugins/PeriodicAllocMonitor.cc

```diff
@@ -0,0 +1,152 @@
+      while (a > max) {
+        if (maxActual_.compare_exchange_strong(max, a, std::memory_order_acq_rel)) {
+          break;
+        }
+      }
+
+      auto single = maxSingleRequested_.load(std::memory_order_relaxed);
+      while (iRequested > single) {
+        if (maxSingleRequested_.compare_exchange_strong(single, iRequested, std::memory_order_acq_rel)) {
+          break;
+        }
+      }
+    }
+    void deallocCalled(size_t iActual) final {
+      if (0 == iActual)
+        return;
+      nDeallocations_.fetch_add(1, std::memory_order_acq_rel);
+      auto present = presentActual_.load(std::memory_order_acquire);
+      if (present >= iActual) {
+        presentActual_.fetch_sub(iActual, std::memory_order_acq_rel);
+      }
+    }
+
+    std::atomic<size_t> requested_ = 0;
+    std::atomic<size_t> presentActual_ = 0;
+    std::atomic<size_t> maxActual_ = 0;
+    std::atomic<size_t> nAllocations_ = 0;
+    std::atomic<size_t> nDeallocations_ = 0;
+    std::atomic<size_t> maxSingleRequested_ = 0;
+  };
+
+}  // namespace
+
+class PeriodicAllocMonitor {
+public:
+  PeriodicAllocMonitor(edm::ParameterSet const& iPS, edm::ActivityRegistry& iAR) {
+    auto adaptor = cms::perftools::AllocMonitorRegistry::instance().createAndRegisterMonitor<MonitorAdaptor>();
+    auto fileName = iPS.getUntrackedParameter<std::string>("filename");
+    auto interval = iPS.getUntrackedParameter<unsigned long long>("millisecondsPerMeasurement");
+
+    thread_ = std::thread([this, fileName, interval, adaptor]() {
+      auto const start = std::chrono::steady_clock::now();
+      std::ofstream fs(fileName);
+      fs << "timestamp, runs-started, lumis-started, events-started, events-finished, total-requested, max-actual, "
+            "present-actual, max-single, nAllocs, nDeallocs\n";
+      while (continueRunning_.load()) {
+        auto rStarted = nRunsStarted_.load(std::memory_order_acquire);
+        auto lStarted = nLumisStarted_.load(std::memory_order_acquire);
+        auto const now = std::chrono::steady_clock::now();
+        auto eStarted = nEventsStarted_.load(std::memory_order_acquire);
+        auto eFinished = nEventsFinished_.load(std::memory_order_acquire);
+        auto report = adaptor->report();
+
+        fs << std::chrono::duration_cast<std::chrono::milliseconds>(now - start).count() << ", " << rStarted << ", "
+           << lStarted << ", " << eStarted << ", " << eFinished << ", " << report.requested_ << ", "
+           << report.maxActual_ << ", " << report.presentActual_ << ", " << report.maxSingleRequested_ << ", "
+           << report.nAllocations_ << ", " << report.nDeallocations_ << std::endl;
+        std::this_thread::sleep_for(std::chrono::milliseconds(interval));
+      }
+    });
+
+    //iAR.postBeginJobSignal_.connect([adaptor]() { adaptor->start(); });
+    iAR.preEventSignal_.connect([this](auto const&) { nEventsStarted_.fetch_add(1, std::memory_order_acq_rel); });
```

Comment:
Could these use the `iAR.watchPreEvent()` interface to register the callbacks? (not a big deal, and I see `EventProcessingAllocMonitor` already slipped through)

Reply:
done